pkg/xds/bootstrap/template_v3.go

```diff
@@ -295,6 +297,34 @@ func genConfig(parameters configParameters) (*envoy_bootstrap_v3.Bootstrap, erro
 	return res, nil
 }
 
+func dnsLookupFamilyFromXdsHost(host string, lookupFn func(host string) ([]net.IP, error)) envoy_cluster_v3.Cluster_DnsLookupFamily {
+	if govalidator.IsHost(host) {
+		ips, err := lookupFn(host)
+		if err != nil {
+			log.Info("[WARNING] error looking up XDS host to determine DnsLookupFamily, falling back to AUTO", "hostname", host)
+			return envoy_cluster_v3.Cluster_AUTO
+		}
+		hasIPv6 := false
+		hasIPv4 := false
+
+		for _, ip := range ips {
+			if ip.To4() != nil {
+				hasIPv4 = true
+			} else {
+				hasIPv6 = true
+			}
+		}
+
+		if hasIPv6 {
+			return envoy_cluster_v3.Cluster_AUTO
+		} else if hasIPv4 {
+			return envoy_cluster_v3.Cluster_V4_ONLY
+		}
```

Comment:
Nit: may it's more readable if you only look at whether there's a no v6 address (all other cases are Cluster_AUTO no?).

Reply:
simplified